taichi/transforms/type_check.cpp

```diff
@@ -20,7 +20,7 @@ class TypeCheck : public IRVisitor {
   TypeCheck(Kernel *kernel) : kernel(kernel) {
     // TODO: remove dependency on get_current_program here
     if (current_program != nullptr)
-      config = get_current_program().config;
+      config = kernel->program.config;
```

Comment:
I am not sure if this is semantically correct, if so, then why it's not written like this before.

Reply:
Because `kernel` might be a `nullptr`? Note that before your changes there's no way to get `Kernel` from IR. So sometimes we have to pass `nullptr` here.

Reply:
Then maybe change `TypeCheck`'s ctor from `TypeCheck(Kernel*)` to `TypeCheck(IRNode*)` would be a way to give a try. Using `IRNode* root` passed to `typecheck`.

Reply:
Feel free to modify the constructors of the classes in the IR passes :)